### PR TITLE
Transforms

### PR DESCRIPTION
**Pull request by mrinalTheCoder**
# Description
This PR does the following:
- Creates ROS2 transform tree anchored at a world frame
- Calculates position of robot `base_link` frame relative to world_frame using GPS and publishes a TF2 frame.
- Calculates waypoint position and publishes a TF2 frame.

# Testing Steps (if relevant)
1. Run bringup
2. Run the `tf.launch.py` file
3. Use the web interface to publish a waypoint/base station (optional)
4. Run `ros2 run tf2_tools view_frames`

Expectation: There should a pdf file with the tf2 tree anchored at world.

# Self Checklist
- [X] I have formatted my code using `ament_uncrustify --reformat`
- [X] I have tested that the new behavior works

### CODE REVIEW

**Comment by mrinalTheCoder** [comment]
tf2 usage in orchestrator

File: urc_navigation/CMakeLists.txt
Comment (yambati03) [inline urc_navigation/CMakeLists.txt:1]: The `urc_navigation` package is going to be completely refactored in another PR, so I would not make any changes to it in this PR.

**Review by yambati03 — CHANGES_REQUESTED** [review]

File: urc_tf/src/world_frame.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+  if (baseStationLat < -200 || baseStationLng < -200) {
+    return;
+  }
+  geometry_msgs::msg::TransformStamped t;
+  t.header.stamp = this->get_clock()->now();
+  t.header.frame_id = "world";
+  t.child_frame_id = "waypoint";
+
+  t.transform.translation.x = (waypointLat - baseStationLat) * 111139;
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:85]: I would swap this literal for a named constant. It's better for readability because `111139` doesn't tell use what this number represents.

**Review by yambati03 — CHANGES_REQUESTED** [review]

**Review by davidcalderon03 — CHANGES_REQUESTED** [review]
Make sure to utilize the GPS const you created, and delete astar and planner_server from urc_orchestrator

File: urc_tf/src/world_frame.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  gps_subscriber = create_subscription<sensor_msgs::msg::NavSatFix>(
+      "/fix", rclcpp::SystemDefaultsQoS(),
+      [this](const sensor_msgs::msg::NavSatFix msg) { GPSCallback(msg); });
+
+  waypoint_subscriber = create_subscription<urc_msgs::msg::Waypoint>(
+      "/waypoint", rclcpp::SystemDefaultsQoS(),
+      [this](const urc_msgs::msg::Waypoint msg) { WaypointCallback(msg); });
+
+  set_base_subscriber = create_subscription<std_msgs::msg::Bool>(
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:16]: Ideally, this shouldn't be subscribing to a topic. Instead, you'd want to expose a service that sets the base coordinates when called. Maybe something like this?
```
float64 latitude
float64 longitude
float64 altitude
---
std_msgs/Empty response
```

File: urc_tf/src/world_frame.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+void WorldFrameBroadcaster::SetBaseCallback(const std_msgs::msg::Bool &msg) {
+  this->baseStationLat = currentLat;
+  this->baseStationLng = currentLng;
+  this->baseStationAlt = currentAlt;
+}
+
+void WorldFrameBroadcaster::IMUCallback(const sensor_msgs::msg::Imu &msg) {
+  this->imu_x = msg.orientation.x;
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:36]: There's no need to unpack all of the IMU message here. Just save the whole thing as a `sensor_msgs::msg::Imu` type and later set `t.transform.rotation = imu_msg.orientation`. This is much cleaner since you never use the individual values!

File: urc_tf/src/world_frame.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  geometry_msgs::msg::TransformStamped t;
+  t.header.stamp = this->get_clock()->now();
+  t.header.frame_id = "odom";
+  t.child_frame_id = "base_link";
+
+  t.transform.translation.x = (currentLat - baseStationLat) * gpsOffsetToMetres;
+  t.transform.translation.y = (currentLng - baseStationLng) * gpsOffsetToMetres;
+  t.transform.translation.z = currentAlt - baseStationAlt;
+
+  t.transform.rotation.x = this->imu_x;
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:60]: Once you address the previous comment, just do `t.transform.rotation = imu_msg.orientation`. Just assuming you call the class variable `imu_msg`.

File: urc_tf/src/world_frame.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      "/imu/data", rclcpp::SystemDefaultsQoS(),
+      [this](const sensor_msgs::msg::Imu msg) { IMUCallback(msg); });
+
+  baseStationLat = -1000;
+  baseStationLng = -1000;
+  baseStationAlt = -1000;
+}
+
+void WorldFrameBroadcaster::SetBaseCallback(const std_msgs::msg::Bool &msg) {
+  this->baseStationLat = currentLat;
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:30]: Did you mean to set this to `currentLat`,  `currentLng`, and `currentAlt`?

File: urc_tf/src/world_frame.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+void WorldFrameBroadcaster::IMUCallback(const sensor_msgs::msg::Imu &msg) {
+  this->imu_x = msg.orientation.x;
+  this->imu_y = msg.orientation.y;
+  this->imu_z = msg.orientation.z;
+  this->imu_w = msg.orientation.w;
+}
+
+void WorldFrameBroadcaster::GPSCallback(
+    const sensor_msgs::msg::NavSatFix &msg) {
+  this->currentLat = msg.latitude;
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:44]: Why do you save these to class variables here?

File: urc_tf/src/world_frame.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  this->currentLng = msg.longitude;
+  this->currentAlt = msg.altitude;
+
+  if (baseStationLat < -200 || baseStationLng < -200) {
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:48]: Also print an error here using `RCLCPP_ERROR(...)`

Comment (yambati03) [reply urc_tf/src/world_frame.cpp:48]: Also, this way of checking if the base station's lat/lng is set is a little awkward. Maybe create a boolean that is set to true when this is set? Then you can just do:
```cpp
if (!isBaseStationCoordinatesSet) {
    return;
}
```

File: urc_tf/src/world_frame.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  t.transform.rotation.y = this->imu_y;
+  t.transform.rotation.z = this->imu_z;
+  t.transform.rotation.w = this->imu_w;
+
+  tf_broadcaster_->sendTransform(t);
+}
+
+void WorldFrameBroadcaster::WaypointCallback(
+    const urc_msgs::msg::Waypoint &msg) {
+  this->waypointLat = msg.latitude;
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:70]: Why is this saved to a class variable? I'd just directly use `msg.latitude` and `msg.longitude` in the function unless you're using it somewhere else.

File: urc_tf/src/world_frame.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  this->waypointLng = msg.longitude;
+
+  if (baseStationLat < -200 || baseStationLng < -200) {
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:73]: See above comment.

File: urc_tf/src/world_frame.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    return;
+  }
+  geometry_msgs::msg::TransformStamped t;
+  t.header.stamp = this->get_clock()->now();
+  t.header.frame_id = "world";
```
Comment (yambati03) [inline urc_tf/src/world_frame.cpp:78]: Just checking, is this supposed to be "world"?

**Review by yambati03 — CHANGES_REQUESTED** [review]
Looks good! Just some comments to refactor for cleanliness/readability.

**Review by davidcalderon03 — APPROVED** [review]
